### ISSUE TITLE
RFC: Deprecate DataParallel2 Strategy (DDP2)

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/lightning#12584]
## Proposed refactor

Deprecate the DDP2 strategy.

### Motivation

We have few to no users. Evidence:

- The strategy hasn't been working for at least two PL versions
- It is broken for all PyTorch versions >= 1.9 and it is unclear how to make it work for PyTorch >= 1.9. [Explanation](https://github.com/PyTorchLightning/pytorch-lightning/issues/9667#issuecomment-986161540)
- We get no questions, bug reports, and no feature requests related to DDP2
- There are no functional tests for DDP2

Hence, it is not worth the maintenance effort to keep DDP2, especially since it is broken.

### Pitch

Deprecate and remove DDP2 in PL 1.9.

### Additional context

Closes issues #9667

Code repository search shows that the only usages are from within PL and forks of PL:
https://grep.app/search?q=DDP2Plugin
https://github.com/search?p=1&q=DDP2Plugin&type=Code

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @tchaton

**Comment by justusschock** [comment]
is there any use case where one wants to use this over ddp? the only one I could think of would be jupyter notebooks with distributed machines in the back. But there ddp_spawn should work as well, right?

so if there is no workflow where *only* ddp2 works, i’m in favor of this

**Comment by carmocca** [comment]
I am also in favor as I have never seen it used.

> It is broken for all PyTorch versions >= 1.9 and it is unclear how to make it work for PyTorch >= 1.9. https://github.com/PyTorchLightning/pytorch-lightning/issues/9667#issuecomment-986161540

PL 1.7 will have PyTorch 1.9 as its minimum version, so do we want to keep the code until 1.9 if it's broken anyway?
We could make this a breaking change with the 1.7 release by it and raising a `ValueError` describing why it has been removed. Then remove this error in 1.9

**Comment by awaelchli** [comment]
> is there any use case where one wants to use this over ddp?


@justusschock The original use case was that DDP2 behaves
1.  Like DP in each node, i.e., the full batch of inputs in a single node gets scattered and the outputs of all gpus in a single node get gathered. This was useful when the loss needed to be defined over the full batch and large batches are needed. 
2. Like DDP across the nodes, i.e., the gradients were synced across all nodes. 

What's important here is that DDP2 does not add any value over DP when training on a single node. It's identical in this case. So DDP2 can be seen as a extension of DP to multi-node. That's what makes it unique and why you'd want to use it over DDP (and in DDP you don't have access to the full batch of all processes). 

A concrete example, I think, would be contrastive learning where one would make a large batch where half of the samples are positive and the other portion negative, and the contrastive loss is defined over the entire output batch. 

Something like this afaik would not be possible with existing strategies. But it seems nobody is asking for this, and if a single node is enough you would just use DP, that's my recommendation. 


> We could make this a breaking change with the 1.7

Yes, probably the best. Better have a good error message than a deprecation + a low-level cryptic error message.

**Comment by thomassajot** [comment]
Hi @awaelchli, 
My team is working on a contrastive learning project. We are trying to use `ddp2` recently in hopes to get larger batch sizes. I found this issue while trying to debug why `ddp2` is not working ! 

I understand that if few users are relying on this strategy, it is not a good case to keep it. But of course, I would prefer to have it working.

**Comment by awaelchli** [comment]
Hi @thomassajot 
Sorry for the late answer. 

> I understand that if few users are relying on this strategy, it is not a good case to keep it

This is not actually the main reason we remove it. Have you seen my [explanation](https://github.com/PyTorchLightning/pytorch-lightning/issues/9667#issuecomment-986161540) of why it doesn't work? 

Essentially you would like to use the DistributedDataParallel wrapper with multiple device_ids but PyTorch discouraged and deprecated this a long time ago. Unfortunately don't currently see a way to work around or keep support for it as we fully rely on PyTorch here for all distributed logic.

Your use case was exactly the same use case for when DDP2 was originally added to PL. For contrastive learning.
